libpam/include/security/pam_sandbox.h

```diff
@@ -0,0 +1,136 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/landlock.h>
+#include <linux/prctl.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/prctl.h>
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#ifndef landlock_create_ruleset
+
+	static inline int
+landlock_create_ruleset(const struct landlock_ruleset_attr *const attr,
+		const size_t size, const __u32 flags)
+{
+	return syscall(__NR_landlock_create_ruleset, attr, size, flags);
+}
+#endif
+
+#ifndef landlock_add_rule
+static inline int landlock_add_rule(const int ruleset_fd,
+		const enum landlock_rule_type rule_type,
+		const void *const rule_attr,
+		const __u32 flags)
+{
+	return syscall(__NR_landlock_add_rule, ruleset_fd, rule_type, rule_attr,
+			flags);
+}
+#endif
+
+#ifndef landlock_restrict_self
+static inline int landlock_restrict_self(const int ruleset_fd,
+		const __u32 flags)
+{
+	return syscall(__NR_landlock_restrict_self, ruleset_fd, flags);
+}
+#endif
+
+#define ACCESS_FS_ROUGHLY_READ ( \
+		LANDLOCK_ACCESS_FS_EXECUTE | \
+		LANDLOCK_ACCESS_FS_READ_FILE | \
+		LANDLOCK_ACCESS_FS_READ_DIR)
+
+#ifdef LANDLOCK_ACCESS_FS_REFER
+#define ACCESS_FS_ROUGHLY_WRITE ( \
```

Comment:
As a general note, we should only have one `ACCESS_FS_ROUGHLY_WRITE`, always including `LANDLOCK_ACCESS_FS_REFER`, but removing it at runtime if required. See https://git.kernel.org/pub/scm/linux/kernel/git/mic/linux.git/commit/?h=next&id=c773739d5362a8a3fc595c9851119562cb26f2c7

However, for this PAM use case, let's make it simple and just remove `LANDLOCK_ACCESS_FS_REFER` everywhere. Indeed, this access right is denied by default even if the ruleset doesn't handle it, and because this sandboxed PAM will not use rename(2) nor link(2), it should be fine.

Reply:
Seem's fair for this case but as the `pam_sandbox.h` header could be used in other modules shouldn't it integrate `LANDLOCK_ACCESS_FS_REF` checks anyway ?